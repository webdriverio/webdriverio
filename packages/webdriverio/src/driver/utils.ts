import cp from 'node:child_process'

import decamelize from 'decamelize'
import logger from '@wdio/logger'
import { getChromePath } from 'chrome-launcher'
import {
    install, canDownload, resolveBuildId, detectBrowserPlatform, Browser, ChromeReleaseChannel,
    computeExecutablePath, type InstallOptions
} from '@puppeteer/browsers'
import type { EdgedriverParameters } from 'edgedriver'
import type { Capabilities } from '@wdio/types'

const log = logger('webdriver')
const EXCLUDED_PARAMS = ['version', 'help']

export function parseParams(params: EdgedriverParameters) {
    return Object.entries(params)
        .filter(([key,]) => !EXCLUDED_PARAMS.includes(key))
        .map(([key, val]) => {
            if (typeof val === 'boolean' && !val) {
                return ''
            }
            const vals = Array.isArray(val) ? val : [val]
            return vals.map((v) => `--${decamelize(key, { separator: '-' })}${typeof v === 'boolean' ? '' : `=${v}`}`)
        })
        .flat()
        .filter(Boolean)
}

export function getLocalChromePath() {
    try {
        const chromePath = getChromePath()
        return chromePath
    } catch (err: unknown) {
        return
    }
}

export function getBuildIdByPath(chromePath?: string) {
    if (!chromePath) {
        return
    }
    const versionString = cp.execSync(`"${chromePath}" --version`).toString()
    return versionString.split(' ').pop()?.trim()
}

export async function setupChrome(caps: Capabilities.Capabilities, cacheDir: string) {
    const platform = detectBrowserPlatform()
    if (!platform) {
        throw new Error('The current platform is not supported.')
    }

    if (!caps.browserVersion) {
        const executablePath = getLocalChromePath()!
        const tag = getBuildIdByPath(executablePath)

        /**
         * verify that we have a valid Chrome browser installed
         */
        if (tag) {
            return {
                platform,
                executablePath,
                buildId: await resolveBuildId(Browser.CHROME, platform, tag)
            }
        }
    }

    /**
     * otherwise download provided Chrome browser version or "stable"
     */
    const tag = caps.browserVersion || ChromeReleaseChannel.STABLE
    const buildId = await resolveBuildId(Browser.CHROME, platform, tag)
    const installOptions: InstallOptions = {
        cacheDir,
        platform,
        buildId,
        browser: Browser.CHROME,
        downloadProgressCallback: () => {}
    }
    const isCombinationAvailable = await canDownload(installOptions)
    if (!isCombinationAvailable) {
        throw new Error(`Couldn't find a matching Chrome browser for tag "${buildId}" on platform "${platform}"`)
    }

    log.info(`Downloading Chrome v${buildId}`)
    await install(installOptions)
    const executablePath = computeExecutablePath(installOptions)
    return { executablePath, buildId, platform }
}
