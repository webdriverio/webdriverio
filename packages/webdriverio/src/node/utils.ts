import path from 'node:path'
import fs from 'node:fs/promises'

/**
 * check if directory exists
 */
export async function assertDirectoryExists(filepath: string) {
    const exist = await fs.access(path.dirname(filepath)).then(() => true, () => false)
    if (!exist) {
        throw new Error(`directory (${path.dirname(filepath)}) doesn't exist`)
    }
}

/**
 * Get absolute file path of a file
 * @param filepath given file path
 * @returns absolute file path
 */
export function getAbsoluteFilepath(filepath: string) {
    return filepath.startsWith('/') || filepath.startsWith('\\') || filepath.match(/^[a-zA-Z]:\\/)
        ? filepath
        : path.join(process.cwd(), filepath)
}
