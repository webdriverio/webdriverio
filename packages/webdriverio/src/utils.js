import fs from 'fs'
import path from 'path'
import cssValue from 'css-value'
import rgb2hex from 'rgb2hex'
import GraphemeSplitter from 'grapheme-splitter'

import { ELEMENT_KEY, W3C_SELECTOR_STRATEGIES, UNICODE_CHARACTERS } from './constants'

const DEFAULT_SELECTOR = 'css selector'
const DIRECT_SELECTOR_REGEXP = /^(id|css selector|xpath|link text|partial link text|name|tag name|class name|-android uiautomator|-ios uiautomation|accessibility id):(.+)/
const INVALID_SELECTOR_ERROR = new Error('selector needs to be typeof `string` or `function`')

export const findStrategy = function (value, isW3C, isMobile) {
    /**
     * set default selector
     */
    let using = DEFAULT_SELECTOR

    /**
     * check if user has specified locator strategy directly
     */
    const match = value.match(DIRECT_SELECTOR_REGEXP)
    if (match) {
        /**
         * ensure selector strategy is supported
         */
        if (!isMobile && isW3C && !W3C_SELECTOR_STRATEGIES.includes(match[1])) {
            throw new Error('InvalidSelectorStrategy') // ToDo: move error to wdio-error package
        }

        return {
            using: match[1],
            value: match[2]
        }
    }

    // use xPath strategy if value starts with //
    if (value.indexOf('/') === 0 || value.indexOf('(') === 0 ||
               value.indexOf('../') === 0 || value.indexOf('./') === 0 ||
               value.indexOf('*/') === 0) {
        using = 'xpath'

    // use link text strategy if value starts with =
    } else if (value.indexOf('=') === 0) {
        using = 'link text'
        value = value.slice(1)

    // use partial link text strategy if value starts with *=
    } else if (value.indexOf('*=') === 0) {
        using = 'partial link text'
        value = value.slice(2)

    // recursive element search using the UiAutomator library (Android only)
    } else if (value.indexOf('android=') === 0) {
        using = '-android uiautomator'
        value = value.slice(8)

    // recursive element search using the UIAutomation library (iOS-only)
    } else if (value.indexOf('ios=') === 0) {
        using = '-ios uiautomation'
        value = value.slice(4)

    // recursive element search using accessibility id
    } else if (value.indexOf('~') === 0) {
        using = 'accessibility id'
        value = value.slice(1)

    // class name mobile selector
    // for iOS = UIA...
    // for Android = android.widget
    } else if (value.slice(0, 3) === 'UIA' || value.slice(0, 15) === 'XCUIElementType' || value.slice(0, 14).toLowerCase() === 'android.widget') {
        using = 'class name'

    // use tag name strategy if value contains a tag
    // e.g. "<div>" or "<div />"
    } else if (value.search(/<[a-zA-Z-]+( \/)*>/g) >= 0) {
        using = 'tag name'
        value = value.replace(/<|>|\/|\s/g, '')

    // use name strategy if value queries elements with name attributes for JSONWP
    // e.g. "[name='myName']" or '[name="myName"]'
    } else if (!isW3C && value.search(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/) >= 0) {
        using = 'name'
        value = value.match(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/)[2]

    // allow to move up to the parent or select current element
    } else if (value === '..' || value === '.') {
        using = 'xpath'

    // any element with given class, id, or attribute and content
    // e.g. h1.header=Welcome or [data-name=table-row]=Item or #content*=Intro
    } else {
        const match = value.match(new RegExp([
            // HTML tag
            /^([a-z0-9|-]*)/,
            // optional . or # + class or id
            /(?:(\.|#)(-?[_a-zA-Z]+[_a-zA-Z0-9-]*))?/,
            // optional [attribute-name="attribute-value"]
            /(?:\[(-?[_a-zA-Z]+[_a-zA-Z0-9-]*)(?:=(?:"|')([a-zA-z0-9\-_. ]+)(?:"|'))?\])?/,
            // *=query or =query
            /(\*)?=(.+)$/,
        ].map(rx => rx.source).join('')))

        if (match) {
            const PREFIX_NAME = { '.': 'class', '#': 'id' }
            const conditions = []
            const [
                tag,
                prefix, name,
                attrName, attrValue,
                partial, query
            ] =  match.slice(1)

            if (prefix) {
                conditions.push(`contains(@${PREFIX_NAME[prefix]}, "${name}")`)
            }
            if (attrName) {
                conditions.push(
                    attrValue
                        ? `contains(@${attrName}, "${attrValue}")`
                        : `@${attrName}`
                )
            }
            if (partial) {
                conditions.push(`contains(., "${query}")`)
            } else {
                conditions.push(`normalize-space() = "${query}"`)
            }

            using = 'xpath'
            value = `.//${tag || '*'}[${conditions.join(' and ')}]`
        }
    }

    /**
     * ensure selector strategy is supported
     */
    if(!isMobile && isW3C && !W3C_SELECTOR_STRATEGIES.includes(using)){
        throw new Error('InvalidSelectorStrategy') // ToDo: move error to wdio-error package
    }

    return { using, value }
}

const applyScopePrototype = (prototype, scope) => {
    const dir = path.resolve(__dirname, 'commands', scope)
    const files = fs.readdirSync(dir)
    for (let filename of files) {
        const commandName = path.basename(filename, path.extname(filename))
        prototype[commandName] = { value: require(path.join(dir, commandName)).default }
    }
}

/**
 * enhances objects with element commands
 */
export const getPrototype = (scope) => {
    const prototype = {}

    /**
     * register action commands
     */
    applyScopePrototype(prototype, scope)
    return prototype
}

/**
 * get element id from WebDriver response
 * @param  {object} res         body object from response
 * @return {string|undefined}   element id or null if element couldn't be found
 */
export const getElementFromResponse = (res) => {
    /**
     * deprecated JSONWireProtocol response
     */
    if (res.ELEMENT) {
        return res.ELEMENT
    }

    /**
     * W3C WebDriver response
     */
    if (res[ELEMENT_KEY]) {
        return res[ELEMENT_KEY]
    }

    return null
}

/**
 * traverse up the scope chain until browser element was reached
 */
export function getBrowserObject (elem) {
    return elem.parent ? getBrowserObject(elem.parent) : elem
}

/**
 * transform whatever value is into an array of char strings
 */
export function transformToCharString (value) {
    const ret = []

    if (!Array.isArray(value)) {
        value = [value]
    }

    for (const val of value) {
        if (typeof val === 'string') {
            ret.push(...checkUnicode(val))
        } else if (typeof val === 'number') {
            const entry = `${val}`.split('')
            ret.push(...entry)
        } else if (val && typeof val === 'object') {
            try {
                ret.push(...JSON.stringify(val).split(''))
            } catch (e) { /* ignore */ }
        } else if (typeof val === 'boolean') {
            const entry = val ? 'true'.split('') : 'false'.split('')
            ret.push(...entry)
        }
    }

    return ret
}

function sanitizeCSS (value) {
    /* istanbul ignore next */
    if (!value) {
        return value
    }

    return value.trim().replace(/'/g, '').replace(/"/g, '').toLowerCase()
}

/**
 * parse css values to a better format
 * @param  {Object} cssPropertyValue result of WebDriver call
 * @param  {String} cssProperty      name of css property to parse
 * @return {Object}                  parsed css property
 */
export function parseCSS (cssPropertyValue, cssProperty) {
    if (!cssPropertyValue) {
        return null
    }

    let parsedValue = {
        property: cssProperty,
        value: cssPropertyValue.toLowerCase().trim()
    }

    if (parsedValue.value.indexOf('rgb') === 0) {
        /**
         * remove whitespaces in rgb values
         */
        parsedValue.value = parsedValue.value.replace(/\s/g, '')

        /**
         * parse color values
         */
        let color = parsedValue.value
        parsedValue.parsed = rgb2hex(parsedValue.value)
        parsedValue.parsed.type = 'color'
        parsedValue.parsed[/[rgba]+/g.exec(color)[0]] = color
    } else if (parsedValue.property === 'font-family') {
        let font = cssValue(cssPropertyValue)
        let string = parsedValue.value
        let value = cssPropertyValue.split(/,/).map(sanitizeCSS)

        parsedValue.value = sanitizeCSS(font[0].value || font[0].string)
        parsedValue.parsed = { value, type: 'font', string }
    } else {
        /**
         * parse other css properties
         */
        try {
            parsedValue.parsed = cssValue(cssPropertyValue)

            if (parsedValue.parsed.length === 1) {
                parsedValue.parsed = parsedValue.parsed[0]
            }

            if (parsedValue.parsed.type && parsedValue.parsed.type === 'number' && parsedValue.parsed.unit === '') {
                parsedValue.value = parsedValue.parsed.value
            }
        } catch (e) {
            // TODO improve css-parse lib to handle properties like
            // `-webkit-animation-timing-function :  cubic-bezier(0.25, 0.1, 0.25, 1)
        }
    }

    return parsedValue
}

/**
 * check for unicode character or split string into literals
 * @param  {String} value  text
 * @return {Array}         set of characters or unicode symbols
 */
export function checkUnicode (value) {
    return UNICODE_CHARACTERS.hasOwnProperty(value) ? [UNICODE_CHARACTERS[value]] : new GraphemeSplitter().splitGraphemes(value)
}

function fetchElementByJSFunction (selector, scope) {
    if (!scope.elementId) {
        return scope.execute(selector)
    }

    const script = ((elem) => (selector).call(elem)).toString().replace('selector', `(${selector.toString()})`)
    return getBrowserObject(scope).execute(`return (${script}).apply(null, arguments)`, scope)
}

/**
 * logic to find an element
 */
export async function findElement(selector) {
    /**
     * fetch element using regular protocol command
     */
    if (typeof selector === 'string') {
        const { using, value } = findStrategy(selector, this.isW3C, this.isMobile)
        return this.elementId
            ? this.findElementFromElement(this.elementId, using, value)
            : this.findElement(using, value)
    }

    /**
     * fetch element with JS function
     */
    if (typeof selector === 'function') {
        const notFoundError = new Error(`Function selector "${selector.toString()}" did not return an HTMLElement`)
        let elem = await fetchElementByJSFunction(selector, this)
        elem = Array.isArray(elem) ? elem[0] : elem
        return getElementFromResponse(elem) ? elem : notFoundError
    }

    throw INVALID_SELECTOR_ERROR
}

/**
 * logic to find a elements
 */
export async function findElements(selector) {
    /**
     * fetch element using regular protocol command
     */
    if (typeof selector === 'string') {
        const { using, value } = findStrategy(selector, this.isW3C, this.isMobile)
        return this.elementId
            ? this.findElementsFromElement(this.elementId, using, value)
            : this.findElements(using, value)
    }

    /**
     * fetch element with JS function
     */
    if (typeof selector === 'function') {
        let elems = await fetchElementByJSFunction(selector, this)
        elems = Array.isArray(elems) ? elems : [elems]
        return elems.filter((elem) => elem && getElementFromResponse(elem))
    }

    throw INVALID_SELECTOR_ERROR
}
