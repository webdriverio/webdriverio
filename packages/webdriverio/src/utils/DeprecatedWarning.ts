import logger from '@wdio/logger'

const log = logger('deprecated')

export class DeprecatedWarning {
    constructor(message: string) {
        log.warn('@DeprecatedWarning: ' + message)
    }
}
