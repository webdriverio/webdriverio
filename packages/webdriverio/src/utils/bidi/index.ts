import { ELEMENT_KEY, type remote, type local } from 'webdriver'

import { EvaluateResultType, NonPrimitiveType, PrimitiveType, RemoteType } from './constants.js'
import { WebdriverBidiExeception } from './error.js'

export function parseScriptResult(params: remote.ScriptCallFunctionParameters, result: local.ScriptEvaluateResult) {
    const type = result.type

    if (type === EvaluateResultType.Success) {
        // @ts-expect-error
        return deserializeValue(result.result)
    }
    if (type === EvaluateResultType.Exception) {
        throw new WebdriverBidiExeception(params, result)
    }

    throw new Error(`Unknown evaluate result type: ${type}`)
}

export function deserializeValue(result: remote.ScriptLocalValue) {
    // @ts-expect-error
    const { type, value } = result
    if (type === NonPrimitiveType.RegularExpression) {
        return new RegExp(value.pattern, value.flags)
    }
    if (type === NonPrimitiveType.Array) {
        return value.map((element: remote.ScriptLocalValue) => deserializeValue(element))
    }
    if (type === NonPrimitiveType.Date) {
        return new Date(value)
    }
    if (type === NonPrimitiveType.Map) {
        return new Map(value.map(([key, value]: [string, remote.ScriptLocalValue]) => (
            [typeof key === 'string' ? key : deserializeValue(key), deserializeValue(value)]
        )))
    }
    if (type === NonPrimitiveType.Set) {
        return new Set(value.map((element: remote.ScriptLocalValue) => deserializeValue(element)))
    }
    if (type === PrimitiveType.Number && value === 'NaN') {
        return NaN
    }
    if (type === PrimitiveType.Number && value === 'Infinity') {
        return Infinity
    }
    if (type === PrimitiveType.Number && value === '-Infinity') {
        return -Infinity
    }
    if (type === PrimitiveType.Number && value === '-0') {
        return -0
    }
    if (type === PrimitiveType.BigInt) {
        return BigInt(value)
    }
    if (type === PrimitiveType.Null) {
        return null
    }
    if (type === NonPrimitiveType.Object) {
        return Object.fromEntries((value || []).map(([key, value]: [string, remote.ScriptLocalValue]) => {
            return [typeof key === 'string' ? key : deserializeValue(key), deserializeValue(value)]
        }))
    }
    if (type === RemoteType.Node) {
        return { [ELEMENT_KEY]: (result as { sharedId: string }).sharedId }
    }
    if (type === RemoteType.Error) {
        return new Error('<unserializable error>')
    }
    return value
}
