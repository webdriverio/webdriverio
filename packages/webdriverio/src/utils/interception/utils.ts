import type { local, remote } from 'webdriver'
import type { RequestWithOptions, RespondWithOptions } from './types.js'

/**
 * parse request or response overwrites to make it compatible with Bidis protocol
 * @param overwrite request or response overwrite
 * @returns object to pass to the protocol
 */
export function parseOverwrite(overwrite: RequestWithOptions | RespondWithOptions, request: local.NetworkBeforeRequestSentParameters | local.NetworkResponseCompletedParameters) {
    let body: remote.NetworkBytesValue | undefined = undefined
    if ('body' in overwrite && overwrite.body) {
        const bodyOverwrite = typeof overwrite.body === 'function'
            ? overwrite.body(request as local.NetworkBeforeRequestSentParameters)
            : overwrite.body
        body = typeof bodyOverwrite === 'string' ?
            /**
             * if body is a string we can pass it as is
             */
            {
                type: 'string',
                value: bodyOverwrite
            }
            :
            /**
             * if body is an object we need to encode it
             */
            {
                type: 'base64',
                value: globalThis.Buffer
                    ? globalThis.Buffer.from(JSON.stringify(bodyOverwrite || '')).toString('base64')
                    : btoa(JSON.stringify(bodyOverwrite))
            }
    }

    let headers: remote.NetworkHeader[] = []
    const headersOverwrite = typeof overwrite.headers === 'function'
        ? overwrite.headers(request as local.NetworkBeforeRequestSentParameters)
        : overwrite.headers
    if (headersOverwrite) {
        headers = Object.entries(headersOverwrite).map(([name, value]) => ({
            name,
            value: { type: 'string', value }
        }))
    }

    let cookies: remote.NetworkSetCookieHeader[] = []
    if ('cookies' in overwrite && overwrite.cookies) {
        const cookieOverwrite = typeof overwrite.cookies === 'function'
            ? overwrite.cookies(request as local.NetworkBeforeRequestSentParameters) || []
            : overwrite.cookies
        cookies = cookieOverwrite.map((cookie) => ({
            name: cookie.name,
            value: <remote.NetworkStringValue>{
                type: 'string',
                value: cookie.value
            },
            domain: cookie.domain,
            path: cookie.path,
            expires: cookie.expiry,
            httpOnly: cookie.httpOnly,
            secure: cookie.secure,
            sameSite: cookie.sameSite?.toLowerCase() as remote.NetworkSetCookieHeader['sameSite'],
        }))
    }

    let statusCode: number | undefined = undefined
    if ('statusCode' in overwrite && overwrite.statusCode) {
        const statusCodeOverwrite = typeof overwrite.statusCode === 'function'
            ? overwrite.statusCode(request as local.NetworkResponseCompletedParameters)
            : overwrite.statusCode
        statusCode = statusCodeOverwrite
    }

    const method: string | undefined = 'method' in overwrite
        ? typeof overwrite.method === 'function'
            ? overwrite.method(request as local.NetworkBeforeRequestSentParameters)
            : overwrite.method
        : undefined

    const url = 'url' in overwrite
        ? typeof overwrite.url === 'function'
            ? overwrite.url(request as local.NetworkBeforeRequestSentParameters)
            : overwrite.url
        : undefined
    return { body, headers, cookies, method, statusCode, url }
}

export function getPatternParam (pattern: URLPattern, key: keyof Omit<remote.NetworkUrlPatternPattern, 'type'>) {
    if (key !== 'pathname' && pattern[key] === '*') {
        return
    }

    if (key === 'port' && pattern.port === '') {
        return pattern.protocol === 'https' ? '443' : '80'
    }

    return pattern[key].replaceAll('*', '\\*')
}
