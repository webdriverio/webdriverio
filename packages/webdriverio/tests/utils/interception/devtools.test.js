import { canAccess } from '@wdio/utils'

import NetworkInterception from '../../../src/utils/interception/devtools'

const cdpClient = {
    send: jest.fn().mockReturnValue(Promise.resolve({
        body: 'eyJmb28iOiJiYXIifQ==',
        base64Encoded: true
    }))
}

beforeEach(() => {
    cdpClient.send.mockClear()
})

test('allows to access network calls', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: []
    })
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/barfoo/test.html' },
        responseHeaders: []
    })
    expect(mock.calls.length).toBe(1)
})

describe('filter by header', () => {
    const mockWithCall = async (filter) => {
        const mock = new NetworkInterception('**/foobar/**', filter)
        await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
            request: { url: 'http://test.com/foobar/test1.html', method: 'put' },
            responseHeaders: [{ name: 'Content-Type', value: 'text/xml' }]
        })
        await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
            request: { url: 'http://test.com/foobar/test2.html', method: 'put' },
            responseHeaders: [{ name: 'Content-Type', value: 'foobar' }]
        })
        return mock
    }

    test('filter network calls by header - 1 match', async () => {
        const mock = await mockWithCall({
            method: 'put',
            headers: { 'Content-Type': 'text/xml' }
        })

        expect(mock.calls.length).toBe(1)
        expect(mock.calls[0].url).toBe('http://test.com/foobar/test1.html')
    })

    test('filter network calls by header - no match value', async () => {
        const mock = await mockWithCall({
            headers: { 'Content-Type': 'no match' }
        })
        expect(mock.calls.length).toBe(0)
    })

    test('filter network calls by header - no match key', async () => {
        const mock = await mockWithCall({
            headers: { 'foo': 'bar' }
        })
        expect(mock.calls.length).toBe(0)
    })
})

describe('filter by postData', () => {
    const mockWithCall = async (filter) => {
        const mock = new NetworkInterception('**/foobar/**', filter)
        await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
            request: { url: 'http://test.com/foobar/test1.html', method: 'post', postData: JSON.stringify({ foo: { bar: 'baz' } }) },
            responseHeaders: []
        })
        await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
            request: { url: 'http://test.com/foobar/test2.html', method: 'post', postData: 'foobar' },
            responseHeaders: []
        })
        await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
            request: { url: 'http://test.com/foobar/test3.html', method: 'get' },
            responseHeaders: []
        })
        return mock
    }

    test('filter network calls by postData - exact match', async () => {
        const mock = await mockWithCall({
            postData: 'foobar'
        })

        expect(mock.calls.length).toBe(1)
        expect(mock.calls[0].url).toBe('http://test.com/foobar/test2.html')
    })

    test('filter network calls by postData - function comparator', async () => {
        const mock = await mockWithCall({
            postData: (postData) => {
                if (typeof postData !== 'string') {
                    return false
                }
                try {
                    return JSON.parse(postData).foo.bar === 'baz'
                } catch {
                    return false
                }
            }
        })

        expect(mock.calls.length).toBe(1)
        expect(mock.calls[0].url).toBe('http://test.com/foobar/test1.html')
    })
})

describe('filter by statusCode', () => {
    const mockWithCall = async (filter) => {
        const mock = new NetworkInterception('**/foobar/**', filter)
        await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
            request: { url: 'http://test.com/foobar/test1.html', method: 'post' },
            responseHeaders: [],
            responseStatusCode: 200
        })
        await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
            request: { url: 'http://test.com/foobar/test2.html', method: 'post' },
            responseHeaders: [],
            responseStatusCode: 201
        })
        return mock
    }

    test('filter network calls by statusCode', async () => {
        const mock = await mockWithCall({
            statusCode: 201
        })

        expect(mock.calls.length).toBe(1)
        expect(mock.calls[0].url).toBe('http://test.com/foobar/test2.html')
    })
})

test('allows to filter network calls by method', async () => {
    const mock = new NetworkInterception('**/foobar/**', {
        method: 'put'
    })
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html', method: 'PUT' },
        responseHeaders: []
    })
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html', method: 'POST' },
        responseHeaders: []
    })
    expect(mock.calls.length).toBe(1)
})

test('decodes base64 responses', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'application/json' }]
    })
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'foobar' }]
    })

    expect(mock.calls[0].body).toEqual({ foo: 'bar' })
    expect(mock.calls[1].body).toBe('{"foo":"bar"}')
})

test('abort request', async () => {
    const request = {
        requestId: 123,
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'application/json' }]
    }
    const mock = new NetworkInterception('**/foobar/**')
    expect(mock.abort.bind(mock)).toThrow()
    expect(() => mock.abort('foobar')).toThrow()

    mock.abortOnce('NameNotResolved')
    mock.abort('ConnectionFailed')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])(request)
    expect(cdpClient.send).toBeCalledWith(
        'Fetch.failRequest',
        {
            requestId: 123,
            errorReason: 'NameNotResolved'
        }
    )

    await NetworkInterception.handleRequestInterception(cdpClient, [mock])(request)
    expect(cdpClient.send).toBeCalledWith(
        'Fetch.failRequest',
        {
            requestId: 123,
            errorReason: 'ConnectionFailed'
        }
    )

    await NetworkInterception.handleRequestInterception(cdpClient, [mock])(request)
    expect(cdpClient.send).toBeCalledWith(
        'Fetch.failRequest',
        {
            requestId: 123,
            errorReason: 'ConnectionFailed'
        }
    )
})

test('stub request with a function', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    mock.respond(() => 'foobar')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        requestId: 123,
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'application/json' }]
    })

    expect(cdpClient.send.mock.calls.pop()).toMatchSnapshot()
})

test('stub request with an object', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    mock.respond({ foo: 'bar' })
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        requestId: 123,
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'application/json' }]
    })

    expect(cdpClient.send.mock.calls.pop()).toMatchSnapshot()
})

test('stub request with a text', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    mock.respond('foobar')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        requestId: 123,
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'application/json' }]
    })

    expect(cdpClient.send.mock.calls.pop()).toMatchSnapshot()
})

test('stub request with a file', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    mock.respond(__filename)
    canAccess.mockImplementation(() => true)
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        requestId: 123,
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'application/json' }]
    })

    const response = cdpClient.send.mock.calls.pop()[1]
    const buff = Buffer.from(response.body, 'base64')
    expect(buff.toString('ascii')).toContain('stub request with a file')
})

test('stub request with a different web resource', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    mock.respond('http://json.org/image.svg')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        requestId: 123,
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [{ name: 'Content-Type', value: 'application/json' }]
    })

    expect(cdpClient.send.mock.calls.pop()[1].responseHeaders).toMatchSnapshot()
})

test('stub request with a different web resource containing different location header', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    mock.respond('http://json.org/image.svg')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        requestId: 123,
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: [
            { name: 'Location', value: 'http://some.other/picture.png' },
            { name: 'Content-Type', value: 'application/json' }
        ]
    })

    expect(cdpClient.send.mock.calls.pop()[1].responseHeaders).toMatchSnapshot()
})

test('allows to clear mocks', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: []
    })
    expect(mock.calls.length).toBe(1)
    mock.clear()
    expect(mock.calls.length).toBe(0)
})

test('allows to restore mocks', async () => {
    const mock = new NetworkInterception('**/foobar/**')
    mock.respondOnce({ foo: 'bar' })
    mock.respond({ bar: 'foo' })

    expect(mock.respondOverwrites.length).toBe(2)
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: []
    })
    expect(mock.respondOverwrites.length).toBe(1)
    await NetworkInterception.handleRequestInterception(cdpClient, [mock])({
        request: { url: 'http://test.com/foobar/test.html' },
        responseHeaders: []
    })
    expect(mock.respondOverwrites.length).toBe(1)

    mock.restore()
    expect(mock.respondOverwrites.length).toBe(0)
})
