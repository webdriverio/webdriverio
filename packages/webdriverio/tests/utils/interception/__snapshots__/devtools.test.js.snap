// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stub request with a different web resource 1`] = `
Array [
  Object {
    "name": "Content-Type",
    "value": "application/json",
  },
  Object {
    "name": "Location",
    "value": "http://json.org/image.svg",
  },
]
`;

exports[`stub request with a different web resource containing different location header 1`] = `
Array [
  Object {
    "name": "Content-Type",
    "value": "application/json",
  },
  Object {
    "name": "Location",
    "value": "http://json.org/image.svg",
  },
]
`;

exports[`stub request with a function 1`] = `
Array [
  "Fetch.fulfillRequest",
  Object {
    "body": "Zm9vYmFy",
    "requestId": 123,
    "responseCode": undefined,
    "responseHeaders": Array [
      Object {
        "name": "Content-Type",
        "value": "application/json",
      },
    ],
  },
]
`;

exports[`stub request with a text 1`] = `
Array [
  "Fetch.fulfillRequest",
  Object {
    "body": "Zm9vYmFy",
    "requestId": 123,
    "responseCode": undefined,
    "responseHeaders": Array [
      Object {
        "name": "Content-Type",
        "value": "application/json",
      },
    ],
  },
]
`;

exports[`stub request with an object 1`] = `
Array [
  "Fetch.fulfillRequest",
  Object {
    "body": "eyJmb28iOiJiYXIifQ==",
    "requestId": 123,
    "responseCode": undefined,
    "responseHeaders": Array [
      Object {
        "name": "Content-Type",
        "value": "application/json",
      },
    ],
  },
]
`;
